src/lib/consensus/proof_of_stake.ml

```diff
@@ -491,48 +491,78 @@ module Vrf = struct
   module Threshold = struct
     open Bignum_bigint
 
-    let of_uint64_exn = Fn.compose of_int64_exn UInt64.to_int64
-
     (* TEMPORARY HACK FOR TESTNETS: c should be 1 otherwise *)
-    let c_int = 2
+    let c : [`One | `Two] = `Two
+
+    let c_int = match c with `One -> 1 | `Two -> 2
 
-    let c = of_int c_int
+    let base = Bignum.(one / of_int 2)
 
-    (*  Check if
-        vrf_output / 2^256 <= c * my_stake / total_currency
+    let params =
+      Snarky_taylor.Exp.params ~base
+        ~field_size_in_bits:Snark_params.Tick.Field.size_in_bits
 
-        So that we don't have to do division we check
+    (* TODO: Make efficient *)
+    let bigint_of_uint64 = Fn.compose Bigint.of_string UInt64.to_string
 
-        vrf_output * total_currency <= c * my_stake * 2^256
+    (*  Check approximately if
+        vrf_output / (c * 2^256) <= 1 - base^(my_stake / total_currency)
     *)
     let is_satisfied ~my_stake ~total_stake vrf_output =
-      of_bit_fold_lsb (Random_oracle.Digest.fold_bits vrf_output)
-      * of_uint64_exn (Amount.to_uint64 total_stake)
-      <= shift_left
-           (c * of_uint64_exn (Balance.to_uint64 my_stake))
-           Random_oracle.Digest.length_in_bits
+      let input =
+        (* get first params.per_term_precision bits of top / bottom.
+
+           This is equal to
+
+           floor(2^params.per_term_precision * top / bottom) / 2^params.per_term_precision
+        *)
+        let k = params.per_term_precision in
+        let top = bigint_of_uint64 (Balance.to_uint64 my_stake) in
+        let bottom = bigint_of_uint64 (Amount.to_uint64 total_stake) in
+        Bignum.(
+          of_bigint Bignum_bigint.(shift_left top k / bottom)
```

Comment:
(prelude: the parameterization of `params.per_term_precision` is unclear to me, so this comment is predicated off of 0 knowledge on what values that can be)

I'm assuming that this shift_left can't overflow. Is there as scenario in which it can overflow? For instance, are you assuming that `top < 2^(bits/2)`, where bits is the number of bits in our field element? If so, you should probably assert that fact since right now we somtimes run testnets with more than 50% stake.